### PR TITLE
Feature/rooms resource

### PR DESCRIPTION
**Pull request by vesuvian**
Got the List Rooms and List Locations API endpoints working

### CODE REVIEW

**Comment by Jericho** [comment]
Thank you for your contribution. I'll review and comment but in the mean time please raise an issue to describe the problem(s) your PR is attempting to solve. This is particularly important since I use a tool to automatically generate release notes and this tool relies on issues to determine the content of the notes.

**Comment by vesuvian** [comment]
Hey Jericho, I'm attempting to get the PR merged into your existing rooms feature branch, I believe this addresses issue https://github.com/Jericho/ZoomNet/issues/28

**Comment by Jericho** [comment]
Ok perfect. I'll link the issue so it gets automatically closed when we merge this PR.

There seems to be a surprising number of modified files though. I was expecting a few new models and modifications to the `Rooms` resource and it's corresponding interface. I'm really surprised you had to modify files like the OAuthTokenHandler, the webhook parser, etc. 

Let me dig into your PR, I'll probably figure out why you made these changes.

**Comment by vesuvian** [comment]
I got a little heavy handed with my change to the response parsing. The rooms response is paginated with token, but it's missing a total count. The parser was expecting a total count, defaulting to 0, but throwing an exception if it couldn't find the property.

I elected to make calls to GetPropertyValue with a default pass false for throwIfMissing.

File: Source/ZoomNet/Extensions/Internal.cs [hunk 1]
```diff
@@ -500,17 +500,9 @@ internal static JToken GetProperty(this JToken item, string name, bool throwIfMi
 			return property;
 		}
 
-		internal static T GetPropertyValue<T>(this JToken item, string name, T defaultValue = default)
+		internal static T GetPropertyValue<T>(this JToken item, string name, T defaultValue = default, bool throwIfMissing = true)
```
Comment (Jericho) [inline Source/ZoomNet/Extensions/Internal.cs:503]: Is this change necessary for this PR? I suspect it isn't so I would prefer to exclude it and submit a separate issue+PR for this change so we can have a dialogue about it without getting in the way of merging the "Rooms" PR.

Besides, this change is the reason that you had to modify files such as the oauth handler and the webhook parser which have nothing to do with the "Rooms" resource so you can back out these changes as well and make this PR much simpler and easier for me to approve.

**Review by Jericho — CHANGES_REQUESTED** [review]
Thanks for working on this. I really appreciate it. Do you intend to implement the other methods as well (e.g.: create a room, delete a room, get room settings, etc.)?

**Comment by Jericho** [comment]
> I got a little heavy handed with my change to the response parsing. The rooms response is paginated with token, but it's missing a total count. The parser was expecting a total count, defaulting to 0, but throwing an exception if it couldn't find the property.

where is it throwing an exception?

```csharp
var totalRecords = jObject.GetPropertyValue<int?>("total_records", null);
```

I was as careful to prevent exceptions when certain values are missing but maybe I missed one?

**Comment by Jericho** [comment]
Please go through the comments in my review and either mark them as "Resolved" if you have addressed them or respond to my comments so we can continue the discussion.

**Comment by Jericho** [comment]
I rebased and resolved the conflicts so you can continue working on this PR.

File: Source/ZoomNet/Models/Room.cs [hunk 2]
```diff
@@ -8,29 +8,35 @@ namespace ZoomNet.Models
 	public class Room
```
[the remark this answers is not in the document]
Comment (vesuvian) [reply Source/ZoomNet/Models/Room.cs:8]: I can add this no problem, but I'm not actually getting activation code in my List Zoom Rooms response

```
{
    "page_size": 30,
    "next_page_token": "",
    "rooms": [
        {
            "id": "xxx",
            "room_id": "xxx",
            "name": "xxx",
            "location_id": "xxx",
            "status": "Offline"
        },
        {
            "id": "xxx",
            "room_id": "xxx",
            "name": "xxx",
            "location_id": "xxx",
            "status": "Available"
        },
        ...
```

[the remark this answers is not in the document]
Comment (Jericho) [reply Source/ZoomNet/Models/Room.cs:8]: The documentation says `Activation Code is the code that is used to complete the setup of the Zoom Room.` so I guess this value is no longer included in the room details after you complete the setup.

Anyway, if you could add this property that would be great.

**Comment by vesuvian** [comment]
@Jericho the number of features in room settings are becoming an organizational problem. Lots of deeply nested objects and the endpoint serves two different response types. I'm working away at adding everything for GetSettingsAsync, which you should be able to see for yourself later today.

How would you like to see UpdateSettingsAsync implemented? Having 50 optional parameters is probably not smart.

https://marketplace.zoom.us/docs/api-reference/zoom-api/rooms/updatezrsettings

**Comment by Jericho** [comment]
> How would you like to see UpdateSettingsAsync implemented? Having 50 optional parameters is probably not smart.

Completely agree that 50 optional parameters would be overwhelming and confusing. 

If I read the documentation correctly, it sounds like room settings are grouped into 5 categories: general room settings, security settings, client alert settings, notification settings and digital signage settings. And also, it sounds like you can update settings in one of these categories without updating settings in another category. So why don't we offer methods to update settings in each of these categories in separate methods. I propose that we add 5 methods like:
- UpdateGeneralSettings
- UpdateSecuritySettings
- UpdateAlertSettings
- UpdateNotificationSettings
- UpdateSignageSettings

Each of these methods would have a much more manageable number of optional parameters. Again, this is assuming I read the documentation correctly. I do not have a Zoom room license so I cannot run tests to validate any of my assumptions, so feel free to correct me if your testing contradicts anything I said.

FYI: I faced a similar situation when designing the method to update a Meeting. The number of parameters was overwhelming and, more importantly, some of these parameters did not apply to some meetings. For example, it didn't make sense to have optional parameters related to a meeting occurrence if a meeting is not recurring. So, I created three methods (see https://github.com/Jericho/ZoomNet/blob/develop/Source/ZoomNet/Resources/Meetings.cs)
- UpdateScheduledMeetingAsync
- UpdateRecurringMeetingAsync
- UpdateMeetingOccurrenceAsync

and I carefully chose the appropriate optional parameters for each of these three scenarios. Conceptually, very similar to the situation you are facing.

**Pull request closed without merging**